util/src/Pos/Util/Wlog.hs

```diff
@@ -1,23 +1,289 @@
+instance MonadState s m => MonadState s (LoggerNameBox m)
+  -- Defined in ‘System.Wlog.LoggerNameBox’
+instance MonadThrow m => MonadThrow (LoggerNameBox m)
+  -- Defined in ‘System.Wlog.LoggerNameBox’
+instance Monad m => HasLoggerName (LoggerNameBox m)
+  -- Defined in ‘System.Wlog.LoggerNameBox’
+-}
+
+{-
+class HasLoggerName (m :: * -> *) where
+  askLoggerName :: m LoggerName
+  default askLoggerName :: (MonadTrans t, t n ~ m, Monad n,
+                            HasLoggerName n) =>
+                           m LoggerName
+  modifyLoggerName :: (LoggerName -> LoggerName) -> m a -> m a
+  default modifyLoggerName :: (mmorph-1.1.2:Control.Monad.Morph.MFunctor
+                                 t,
+                               t n ~ m, Monad n, HasLoggerName n) =>
+                              (LoggerName -> LoggerName) -> m a -> m a
+  	-- Defined in ‘System.Wlog.HasLoggerName’
+instance Monad m => HasLoggerName (NamedPureLogger m)
+  -- Defined in ‘System.Wlog.PureLogging’
+instance (Monad m, HasLoggerName m) => HasLoggerName (StateT a m)
+  -- Defined in ‘System.Wlog.HasLoggerName’
+instance (Monad m, HasLoggerName m) => HasLoggerName (ReaderT a m)
+  -- Defined in ‘System.Wlog.HasLoggerName’
+instance HasLoggerName Identity
+  -- Defined in ‘System.Wlog.HasLoggerName’
+instance (Monad m, HasLoggerName m) => HasLoggerName (ExceptT e m)
+  -- Defined in ‘System.Wlog.HasLoggerName’
+instance Monad m => HasLoggerName (LoggerNameBox m)
+  -- Defined in ‘System.Wlog.LoggerNameBox’
+-}
+
+{-
+launchNamedPureLog ::
+  (System.Wlog.CanLog.WithLogger n, Monad m) =>
+  (forall (f :: * -> *). Functor f => m (f a) -> n (f b))
+  -> NamedPureLogger m a -> n b
+  	-- Defined in ‘System.Wlog.PureLogging’
+-}
+launchNamedPureLog
+    :: (WithLogger n, Monad m)
+    => (forall f. Functor f => m (f a) -> n (f b))
+    -> NamedPureLogger m a
+    -> n b
+launchNamedPureLog hoist' namedPureLogger = do
+    name <- askLoggerName
+    (logs, res) <- hoist' $ swap <$> usingNamedPureLogger name namedPureLogger
+    res <$ dispatchEvents logs
+
+usingNamedPureLogger :: Functor m
+                     => LoggerName
+                     -> NamedPureLogger m a
+                     -> m (a, [LogEvent])
+usingNamedPureLogger name (NamedPureLogger action) =
+    usingLoggerName name $ runPureLog action
+
+data LogEvent = LogEvent
+    { leLoggerName :: !LoggerName
+    , leSeverity   :: !Severity
+    , leMessage    :: !Text
+    } deriving (Show)
+
+runPureLog :: Functor m => PureLogger m a -> m (a, [LogEvent])
+runPureLog = fmap (second toList) . usingStateT mempty . runPureLogger
+
+-- |
+newtype PureLogger m a = PureLogger
+    { runPureLogger :: StateT (Seq LogEvent) m a
+    } deriving (Functor, Applicative, Monad, MonadTrans, MonadState (Seq LogEvent),
+                MonadThrow, HasLoggerName)
+
+instance Monad m => CanLog (PureLogger m) where
+    dispatchMessage leLoggerName leSeverity leMessage = modify' (|> LogEvent{..})
```

Comment:
This is the one use of RWC in this file. You should also be aware that if we enable `-Wall` we immediately have a `-Wshadowed-name` warning there, in that the names there will shadow the names in the `LogEvent` struct defined in this file.

Instead I would write that as:
```
instance Monad m => CanLog (PureLogger m) where
    dispatchMessage name severity message = modify' (|> (LogEvent name severity message))
```
No name shadowing and no RWC.

Reply:
Done.

Reply:
Thanks @erikd , this really makes sense. I was just copy-pasting this to please the compiler; should have also thought more about it.